### PR TITLE
First pass on implementing gamepad trigger press point and deadzone settings

### PR DESCRIPTION
**Pull request by jimon**

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs [hunk 1]
```diff
@@ -849,8 +897,29 @@ private static float GetInputSystemDefaultDeadZone()
...
+                case GamepadSlot.Slot4:
+                case GamepadSlot.Slot5:
+                case GamepadSlot.Slot6:
+                case GamepadSlot.Slot7:
+                case GamepadSlot.Slot8:
+                case GamepadSlot.Slot9:
+                case GamepadSlot.Slot10:
+                case GamepadSlot.Slot11:
+                case GamepadSlot.Slot12:
+                    return s_GamepadConfigs[(int)gamepadSlot].deadzone;
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: I still do not see what we gain here by having GamepadSlot as an enum compared to plain integer? The only special thing seems to be "All" which could be a constant used as default argument? basically the enum prevents any kind of loop/enumeration processing on top which is a likely use-case?

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs [hunk 2]
```diff
@@ -948,7 +1057,32 @@ public static bool IsGamepadButtonUp(GamepadButton input, GamepadSlot gamepadSlo
         /// </remarks>
         public static void SetGamepadTriggerPressPoint(float pressPoint, GamepadSlot gamepadSlot = GamepadSlot.All)
         {
-            throw new NotImplementedException();
+            // TODO remove me when init is implemented
+            InitializeGamepadConfigs();
+
+            switch (gamepadSlot)
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/API/Input.cs:1028]: Similar comment as above to slot enum vs integer

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs [hunk 3]
```diff
@@ -849,8 +897,29 @@ private static float GetInputSystemDefaultDeadZone()
...
+                case GamepadSlot.Slot6:
+                case GamepadSlot.Slot7:
+                case GamepadSlot.Slot8:
+                case GamepadSlot.Slot9:
+                case GamepadSlot.Slot10:
+                case GamepadSlot.Slot11:
+                case GamepadSlot.Slot12:
+                    return s_GamepadConfigs[(int)gamepadSlot].deadzone;
+                case GamepadSlot.All:
+                    return s_GamepadConfigs[0].deadzone; // TODO what should it mean?
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Indeed, feels like its not applicable if this is an aggregated gamepad shouldn't each individual be used for each individual gamepad and aggregation be on top of those processed results?

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs [hunk 4]
```diff
@@ -937,6 +1006,46 @@ public static bool IsGamepadButtonUp(GamepadButton input, GamepadSlot gamepadSlo
         {
             throw new NotImplementedException();
         }
+#endif
+        
+        private static GamepadSlot FindSlotForGamepad(Gamepad gamepad)
+        {
+            // TODO implement me
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Waiting for "stable arrays" for this one?

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs [hunk 5]
```diff
@@ -937,6 +1006,46 @@ public static bool IsGamepadButtonUp(GamepadButton input, GamepadSlot gamepadSlo
...
+                case GamepadSlot.Slot6:
+                case GamepadSlot.Slot7:
+                case GamepadSlot.Slot8:
+                case GamepadSlot.Slot9:
+                case GamepadSlot.Slot10:
+                case GamepadSlot.Slot11:
+                case GamepadSlot.Slot12:
+                    return s_GamepadConfigs[(int)gamepadSlot].triggerPresspoint;
+                case GamepadSlot.All:
+                    return s_GamepadConfigs[0].triggerPresspoint; // TODO what should it mean?
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Agree, similar comment as previously, aggregating this one e.g. min/max or similar would yield a different result than applying it on each individual contributor.

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs [hunk 6]
```diff
@@ -804,7 +847,12 @@ private static Vector2 NormalizeAxis(Vector2 axis, float deadzone)
         public static Vector2 GetAxis(Inputs left, Inputs right, Inputs up, Inputs down)
         {
             var axis = GetAxisRaw(left, right, up, down);
-            var deadzone = GetInputSystemDefaultDeadZone(); // TODO account for controls being on gamepads
+            var deadzone = (GetDeviceTypeForInput(left) == InputDeviceType.Gamepad
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Does this makes sense? It seems like its intended to get dead zone from gamepad slot, but Inputs given above might come from e.g. 4 different gamepads? what dead zone should be used? Feels like deadzone should be passed to this function to move that decision higher up or have some restriction in place?

**Review by ekcoh — COMMENTED** [review]
Providing some initial comments but will have another look when I can digest the full file (not only diffs from this iteration). I will come back to switch-cases and this Inputs enum which currently seem to me like its working against a straight forward implementation.

File: Assets/Tests/InputSystem/CoreTests_HighLevelAPI.cs [hunk 7]
```diff
@@ -185,4 +185,41 @@ public void HighLevelAPI_CanQueryGetAxis()
...
+        gamepadState.leftTrigger = 0.3f;
+        InputSystem.QueueStateEvent(gamepad, gamepadState);
+        InputSystem.Update();
+        
+        Input.SetGamepadTriggerPressPoint(0.5f);
+        Assert.That(Input.IsControlPressed(Inputs.Gamepad_LeftTrigger), Is.False);
+
+        Input.SetGamepadTriggerPressPoint(0.3f);
+        Assert.That(Input.IsControlPressed(Inputs.Gamepad_LeftTrigger), Is.True);
+    }
```
Comment (jamesmcgill) [inline Assets/Tests/InputSystem/CoreTests_HighLevelAPI.cs:210]: I would add a couple of extra tests just to check that if you don't touch `Input.SetGamepadTriggerPressPoint()` again, but instead inject some more `InputSystem.QueueStateEvent()/InputSystem.Update();` call that the pressed state changes that way too.

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs [hunk 8]
```diff
@@ -270,6 +270,37 @@ public static class Input
 
 #endif
 
+        private struct GamepadConfig
+        {
+            public float triggerPresspoint;
```
Comment (jamesmcgill) [inline Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: I feel this should be `PressPoint` with extra capital P. But that is debatable.

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 6 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Review fix: shouldn't be any deadzones here

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 3 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Review fix: shouldn't be all, and passing Any should throw an exception

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs [hunk 9]
```diff
@@ -948,7 +1057,32 @@ public static bool IsGamepadButtonUp(GamepadButton input, GamepadSlot gamepadSlo
...
+                case GamepadSlot.Slot6:
+                case GamepadSlot.Slot7:
+                case GamepadSlot.Slot8:
+                case GamepadSlot.Slot9:
+                case GamepadSlot.Slot10:
+                case GamepadSlot.Slot11:
+                case GamepadSlot.Slot12:
+                    s_GamepadConfigs[(int)gamepadSlot].triggerPresspoint = pressPoint;
+                    break;
+                case GamepadSlot.All:
```
Comment (jimon) [inline Packages/com.unity.inputsystem/InputSystem/API/Input.cs:1044]: Remove Any or All, flip a coin :)

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 1 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: I think doing enums is fine for now, let's check that on first customer API feedback and if int's are preferred then switch to int's

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 3 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Fixed: changed method to throw exception if All is passed

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 4 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Yes, some methods are waiting for stable arrays.

But this one is redundant now, so fixed.

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 5 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Fixed similar to previous method - we throw an exception if all is passed

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 6 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: Fixed by removing gamepad deadzone here

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 9 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs:1044]: Removed .Any

**Review by andrew-oc — APPROVED** [review]

**Review by jamesmcgill — APPROVED** [review]

File: Packages/com.unity.inputsystem/InputSystem/API/Input.cs (hunk 8 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/API/Input.cs]: fixed

File: Assets/Tests/InputSystem/CoreTests_HighLevelAPI.cs (hunk 7 above)
Comment (jimon) [reply Assets/Tests/InputSystem/CoreTests_HighLevelAPI.cs:210]: fixed

**Pull request merged**